src/transaction/flashback.c

```diff
@@ -0,0 +1,418 @@
+ * context (in/out)      : flashback summary context that contains summary list
+ */
+
+void
+flashback_cleanup (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_CONTEXT * context)
+{
+  /* *INDENT-OFF* */
+  for (auto iter:context->summary_list)
+    {
+      if (iter.second != NULL)
+        {
+          flashback_drop_summary_entry (thread_p, iter.second);
+        }
+    }
+  /* *INDENT-ON* */
+}
+
+/*
+ * flashback_make_summary - read log volumes within the range of start_lsa and end_lsa then make summary information
+ *
+ * return                : error code
+ * thread_p (in)         : thread entry
+ * context (in/out)      : flashback summary context that contains range of travering logs, filter (user, class), and summary list
+ */
+
+int
+flashback_make_summary_list (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_CONTEXT * context)
+{
+  LOG_PAGE *log_page_p = NULL;
+  char log_pgbuf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+
+  LOG_LSA cur_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA next_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA process_lsa = LSA_INITIALIZER;
+
+  LOG_RECORD_HEADER *log_rec_header;
+
+  int trid;
+
+  time_t current_time = 0;
+
+  FLASHBACK_SUMMARY_ENTRY *summary_entry = NULL;
+
+  int error = NO_ERROR;
+
+  LOG_RECTYPE log_type;
+
+  char *supplement_data = NULL;
+  int supplement_alloc_length = 0;
+
+  log_page_p = (LOG_PAGE *) PTR_ALIGN (log_pgbuf, MAX_ALIGNMENT);
+
+  assert (!LSA_ISNULL (&context->end_lsa));
+
+  LSA_COPY (&process_lsa, &(context->end_lsa));
+
+  /*fetch log page */
+  if (logpb_fetch_page (thread_p, &process_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+    {
+      goto exit;
+    }
+
+  while (LSA_GE (&process_lsa, &context->start_lsa))
+    {
+      LSA_COPY (&cur_log_rec_lsa, &process_lsa);
+      log_rec_header = LOG_GET_LOG_RECORD_HEADER (log_page_p, &process_lsa);
+
+      log_type = log_rec_header->type;
+      trid = log_rec_header->trid;
+
+      LSA_COPY (&next_log_rec_lsa, &log_rec_header->back_lsa);
+
+      if (LSA_ISNULL (&log_rec_header->prev_tranlsa))
+	{
+	  FLASHBACK_CHECK_AND_GET_SUMMARY (context->summary_list, trid, summary_entry);
+	  if (summary_entry != NULL)
+	    {
+	      /* current time is updated whenever a log record with a time value is encountered.
+	       * this function reads the log records in reverse time order,
+	       * so the current_time represents a time in the future than the log being read now
+	       * Since LOG_DUMMY_HA_SERVER_STATE is logged every second, start_time is set to 1 second before the current_time */
+
+	      summary_entry->start_time = current_time - 1;
+	      LSA_COPY (&summary_entry->start_lsa, &cur_log_rec_lsa);
+	    }
+	}
+
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*log_rec_header), &process_lsa, log_page_p);
+
+      switch (log_type)
+	{
+	case LOG_COMMIT:
+	  {
+	    LOG_REC_DONETIME *donetime;
+
+	    if (context->num_summary == FLASHBACK_MAX_SUMMARY)
+	      {
+		error = ER_FLASHBACK_TOO_MANY_SUMMARY;
```

Comment:
`er_set()` isn't needed?

Reply:
Why don't you use er_set() now?

Reply:
I planned to add er_set() when error message for the error code is defined.